stdlib: Add backward deployment versions for the dynamic-replacement runtime functions

aschwaighofer opened the pull request:
The recent change of how we do dynamic replacements added 2 new runtime
functions. This patch adds those functions to the Compatibility50 static
archive.

This will allow backward deployment to a swift 5.0 runtime.

Patch by Erik Eckstein with a modification to call the standard
libraries implementation (marked as weak) when it is available.

This ensures we can change the implementation in the future and are not
ABI locked.

rdar://problem/51601233

aschwaighofer committed:
stdlib: Add backward deployment versions for the
 dynamic-replacement runtime functions.

The recent change of how we do dynamic replacements added 2 new runtime
functions. This patch adds those functions to the Compatibility50 static
archive.

This will allow backward deployment to a swift 5.0 runtime.

Patch by Erik Eckstein with a modification to call the standard
libraries implementation (marked as weak) when it is available.

This ensures we can change the implementation in the future and are not
ABI locked.

rdar://problem/51601233

aschwaighofer commented:
@jordan I have not moved the functions out into a separate library because we always link in the compatibility library.

https://github.com/apple/swift/blob/master/lib/IRGen/GenDecl.cpp#L453
```
 453   // FIXME: It'd be better to have the driver invocation or build system that
 454   // executes the linker introduce these compatibility libraries, since at
 455   // that point we know whether we're building an executable, which is the only
 456   // place where the compatibility libraries take effect. For the benefit of
 457   // build systems that build Swift code, but don't use Swift to drive
 458   // the linker, we can also use autolinking to pull in the compatibility
 459   // libraries. This may however cause the library to get pulled in in
 460   // situations where it isn't useful, such as for dylibs, though this is
 461   // harmless aside from code size.
 462   if (!IRGen.Opts.UseJIT) {
 463     if (auto compatibilityVersion
 464           = IRGen.Opts.AutolinkRuntimeCompatibilityLibraryVersion) {
 465       if (*compatibilityVersion <= llvm::VersionTuple(5, 0)) {
 466         this->addLinkLibrary(LinkLibrary("swiftCompatibility50",
 467                                          LibraryKind::Library,
 468                                          /*forceLoad*/ true));
 469       }
 470     }
 471   }
 472 }
```

aschwaighofer commented:
@swift-ci please test

jrose-apple commented:
Ah, I didn't realize Joe changed that. Okay.

jckarter commented on include/swift/Runtime/Exclusivity.h:
```diff
@@ -44,13 +44,19 @@ void swift_beginAccess(void *pointer, ValueBuffer *buffer,
 /// be called.
+#ifdef __APPLE__
+__attribute__((weak))
+#endif
```
This will export these symbols as a weak definition, which is probably not what you want. Did you mean `__attribute__((weak_import))` ?

aschwaighofer replied:
Yes. Will fix.

jckarter commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -1225,12 +1225,26 @@ FUNCTION(EndAccess, swift_endAccess, C_CC, AlwaysAvailable,
+         ARGS(FunctionPtrTy->getPointerTo()),
+         ATTRS(NoUnwind))
+
+FUNCTION(GetReplacement50, swift_getFunctionReplacement50, C_CC,
```
I had been thinking about this issue of backporting new entry points with the compatibility library, and I wonder whether there might be a more systematic way of going about it. I was thinking that maybe we could turn the return type of the availability function here from a `bool` to a tristate, `ConditionallyAvailable` | `AvailableByCompatibilityLibrary` | `AlwaysAvailable`. We could factor the logic for choosing the compatibility library or OS version of the entry point into `getRuntimeFn` then, instead of repeating this conditional logic for every client of the entry point.

aschwaighofer committed:
Use weak_import not weak for the optional symbols

aschwaighofer committed:
Refactor to use RuntimeAvailability

jckarter commented:
> Ah, I didn't realize Joe changed that. Okay.

That’s a bug, though, not a feature. The compatibility library *should* only link into the main executable, and dylibs should get back deployment support from the executable. An approach that should work without a separate library, and without relying on the compatibility library getting linked into dylibs, might be to add the back-deployment functions to the static archive as exported entry points, then emit a common resolution thunk that tries to find the entry point, first by looking in libswiftCore.dylib, then by looking in the main executable.

aschwaighofer commented:
@swift-ci Please test

aschwaighofer committed:
Fix MicroStdlib test

aschwaighofer commented:
@swift-ci Please test

aschwaighofer committed:
Really fix micro stdlib test case

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@jckarter I think I have addressed your concerns. Are you okay with the latest version?

jckarter commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -638,7 +638,7 @@ FUNCTION(GetGenericMetadata, swift_getGenericMetadata,
 //                                     uintptr_t index);
 FUNCTION(GetOpaqueTypeMetadata, swift_getOpaqueTypeMetadata,
-         SwiftCC, OpaqueTypeAvailability,
+         SwiftCC, ConditionallyAvailable,
```
These all still need to call a function to return the availability appropriate for the current deployment target. If we're deploying to an OS that always has Swift 5.1, these can all be `AlwaysAvailable`.

aschwaighofer replied:
This is what conditionally available means?

The code in getRuntimeFn() queries whether the feature is available based on whether something is AlwaysAvailable| ConditionallyAvailable | AvailableByCompatibilityLibrary:

```
  auto isFeatureAvailable = [&]() -> bool {
     auto deploymentAvailability =
         AvailabilityContext::forDeploymentTarget(*context);
     auto featureAvailability =();
     return deploymentAvailability.isContainedIn(featureAvailability);
   };

  switch (availability) {
   case RuntimeAvailability:: AlwaysAvailable:
     // Nothing to do.
     break;
   case RuntimeAvailability::ConditionallyAvailable: {
     isWeakLinked = !isFeatureAvailable();
     break;
   }
   case RuntimeAvailability::AvailableByCompatibilityLibrary: {
     if (!isFeatureAvailable())
       functionName.append("50");
     break;
   }
   }
```

In the future we can extend this by adding another entry to runtime.def that specifies at which swift version things became available ...

```
FUNCTION(GetOpaqueTypeMetadata, swift_getOpaqueTypeMetadata,
         SwiftCC, ConditionallyAvailable, Swift51
```


And then generate the availability based on that.
```
 context->get##AVAILVERSION##Availability
```

jckarter replied:
An entry point is only `ConditionallyAvailable` if we're deploying back to a runtime version that may not have the entry point. If we're deploying to an OS with Swift 5.1, everything is `AlwaysAvailable`. Having a single version doesn't seem sufficient to me, because in the future, we may end up in a situation where a runtime entry point can only be back-deployed in the compatibility library to certain runtime versions; for instance, a Swift 5.2 feature may be completely unavailable on Swift 5.0, although we have a compatibility fallback that works for the 5.1 runtime. The most future-proof thing to do is to have the FUNCTION xmacro refer to a function that returns the enum value appropriate to the deployment target.

aschwaighofer replied:
I will address this in a follow-up patch.

aschwaighofer replied:
Attempt of follow-up here: https://github.com/apple/swift/pull/25573

jckarter reviewed (changes requested):
There's still the issue of dynamic libraries. `Compatibility50` will not be linked into dylibs, so they need to get the back-deployed entry points from the one linked into the main executable, if not from the OS runtime.

aschwaighofer committed:
Use thread private key to avoid weak linkage

We use one bit of the third reserved swift private tls key.

Also move the functionality into a separate static archive that is
always linked dependent on deployment target.

aschwaighofer commented:
@swift-ci Please test

jckarter reviewed (approved):
We can address some of these issues in follow up commits.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -904,4 +904,10 @@ def disable_autolinking_runtime_compatibility : Flag<["-"], "disable-autolinking
+    : Flag<[ "-" ],
+    "disable-autolinking-runtime-compatibility-dynamic-replacements">,
+      Flags<[ FrontendOption ]>,
+      HelpText<"Do not use autolinking for runtime compatibility libraries">;
```
Please fix the formatting here to match the other lines, and please change the help text.

…does this really need a separate flag, though?

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -446,6 +446,12 @@ ToolChain::constructInvocation(const CompileJobAction &job,
+  if (context.Args.hasArg(options::
+          OPT_disable_autolinking_runtime_compatibility_dynamic_replacements)) {
+    Arguments.push_back(
+        "-disable-autolinking-runtime-compatibility-dynamic-replacements");
```
This is just `context.Args.AddLastArg(Arguments, OPT_disable_autolinking_runtime_compatibility_dynamic_replacements)` (and Joe should have used that too above).

jrose-apple commented on stdlib/public/runtime/ThreadLocalStorage.h:
```diff
@@ -53,9 +53,14 @@ extern "C" int pthread_key_init_np(int key, void (*destructor)(void *));
 
 # define SWIFT_RUNTIME_TLS_KEY __PTK_FRAMEWORK_SWIFT_KEY0
 # define SWIFT_STDLIB_TLS_KEY __PTK_FRAMEWORK_SWIFT_KEY1
+# define SWIFT_RUNTIME2_TLS_KEY __PTK_FRAMEWORK_SWIFT_KEY2
```
Please give this a more descriptive name, like `SWIFT_COMPATIBILITY_50_TLS_KEY`. That'll indicate when it's okay to reuse it.

aschwaighofer committed:
Fix swiftpm build

aschwaighofer committed:
Address review comments

aschwaighofer committed:
Another review comment I missed

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swift-ci Please test source compatibility

aschwaighofer committed:
Fix invalid deref

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swit-ci Please test source compatibility

aschwaighofer commented:
@swift-ci Please test source compatibility

aschwaighofer committed:
Fix swiftpm targets without swift

SwiftPM will pass a -runtime-compatibility-version none in this case.

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swift-ci Please test source compatibility

aschwaighofer commented:
@swift-ci Please test linux

jrose-apple commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -427,6 +427,22 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
     }
   }
     
+  if (job.getKind() == LinkKind::Executable) {
```
Er, you don't need this part anymore, right? And you never wanted it to just be Executables-linked-with-swiftc anyway.

jckarter replied:
We still ultimately want this to only be linked into executables. We were going to handle that as a followup. For these specific backfill entry points, they're small enough that burning them into dylibs and having potentially-mismatching versions in different images is probably OK, but that's not generally a situation we want to be in.

jrose-apple replied:
I don't understand that part. We can't rely on executables having *any* Swift in them, and it's unlikely that Xcode will switch to linking with swiftc anyway. The autolinking part for dynamic replacement isn't going away.

jrose-apple replied:
(I do want a resolution for this because it's interfering with me rebasing #24787.)

jckarter replied:
If there's any Swift in the process, then the executable is the only useful place to put backward-compatibility fills, so much like arclite, we might want eventually for Xcode to link this in unconditionally because a non-Swift executable could be using dylibs with Swift in them.
